File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
from collections import deque

dx, dy = [1, -1, 0, 0], [0, 0, 1, -1]

T = int(sys.stdin.readline())

def bfs(x, y):
    q = deque()
    q.append((x, y))
    farm[x][y] = 0

    while q:
        y, x = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or nx > M - 1 or ny < 0 or ny > N-1:
                continue

            if farm[ny][nx] == 1:
                farm[ny][nx] = 0
                q.append((ny, nx))

for i in range(T):
    cnt = 0
    M, N, K = map(int, sys.stdin.readline().split())
    farm = [[0]*M for _ in range(N)]
    for i in range(K):
        y, x = map(int, sys.stdin.readline().split())
        farm[x][y] = 1

    for i in range(N):
        for j in range(M):
            if farm[i][j] == 1:
                bfs(i, j)
                cnt += 1
    print(cnt)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {
    public static boolean visited[][];
    public static int map[][];
    public static int dx[];
    public static int dy[];
    public static int M;
    public static int N;
    public static int K;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int T = Integer.parseInt(br.readLine());
        dx = new int[] { 1, 0, -1, 0 };
        dy = new int[] { 0, 1, 0, -1 };

        for (int i = 0; i < T; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            M = Integer.parseInt(st.nextToken());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());
            int cnt = 0;
            map = new int[N][M];
            visited = new boolean[N][M];

            for (int j = 0; j < K; j++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                map[y][x] = 1;
            }

            for (int j = 0; j < N; j++) {
                for (int j2 = 0; j2 < M; j2++) {
                    if (map[j][j2] == 1 && !visited[j][j2]) {
                        bfs(new Point(j2, j));
                        cnt++;
                    }
                }
            }
            sb.append(cnt);
            sb.append("\n");
        }
        System.out.println(sb.toString());
    }

    public static void bfs(Point p) {
        Deque<Point> q = new ArrayDeque<>();
        q.add(p);

        while (!q.isEmpty()) {
            Point pos = q.poll();
            int x = pos.getX();
            int y = pos.getY();
            visited[y][x] = true;

            for (int i = 0; i < 4; i++) {
                int newX = x + dx[i];
                int newY = y + dy[i];
                if (newX < 0 || newX >= M || newY < 0 || newY >= N) {
                    continue;
                }
                if (map[newY][newX] == 1 && !visited[newY][newX]) {
                    q.add(new Point(newX, newY));
                    visited[newY][newX] = true;
                }
            }
        }
    }
}

class Point {
    private int x;
    private int y;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    public int getX() {
        return this.x;
    }

    public int getY() {
        return this.y;
    }
}
